### ISSUE TITLE
CA2000 should be emitted if the result of CancellationToken.Register is not disposed

### ISSUE THREAD

**Issue by SergeyTeplyakov** [dotnet/roslyn-analyzers#4883]
### Analyzer

**Diagnostic ID**: [CA2000](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2000):

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

Event though the result of `CancellationToken.Register` is a disposable struct the CA2000 rule does not emit a diagnostic in the following case:

```csharp
void FooBar(CancellationToken token) {
   token.Register(() => {});
}
```

I noticed that in #1057 there is kind of an explanation why there are no CA2000 diagnostics in the code above, but I don't think that the explanation is sound.

Here is a more detailed example, when the lack of disposal of `CancellationTokenRegistration` actually causes a memory leak:

```csharp
public class VeryExpensiveClass
{
    private readonly string[] _s = new string[10_000];
    public int X { get; set; }
}
public class Service : IDisposable
{
    private readonly CancellationTokenSource _disposeCts = new CancellationTokenSource();

    public async Task LongRunningOperationAsync()
    {
        var veryExpensive = new VeryExpensiveClass();
        await Task.Yield();
        _disposeCts.Token.Register(() => veryExpensive.X++);
    }

    public void Dispose() => _disposeCts.Cancel();
}
```

I.e. if the `CancellationToken` instance is long lived, and the method that calls `CancellationToken.Register` is called a lot, an application will effectively have a memory leak when a long-lived object will have a bunch of references to delegates and (what is more important) to potentially expensive transient objects created in the method.

### Expected behavior

The diagnostics should be emitted when `Register` is called on a non-locally created `CancellationToken`.

### Actual behavior

No diagnostics.

## Additional context

Even though I've found a few cases when a lack of disposal on `RegistrationTokenSource` was causing major memory leaks in our applications, I do understand that this situation is quite tricky.

First, we can't just emit diagnostics all the time. If the `CancellationTokenSource` instance is created locally then it doesn't make any sense to emit this diagnostic.

Second, enforcing this rule may cause a huge pain to the existing users, because in any significant codebase you may get hundreds of diagnostics and only some of them will be actionable. For instance, it's not clear whether to emit the diagnostic here:

```csharp
public static void FooBar(CancellationToken token) {
   token.Register(() => {});
}
```

Maybe the token is created from a short-lived local variable and lack of disposal of `CancellationTokenRegistration` is ok.

**Comment by mavasani** [comment]
We need to add to the set of factory methods/special methods whose return values are tracked by dispose analysis.

**Comment by SergeyTeplyakov** [comment]
Thanks a lot, @mavasani for looking into this.
